Add option to pass object with defaults to environment plugin

einarlove opened the pull request:
**Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [x] Feature

**What is the current behavior?** (You can also link to an open issue here)
You can pass an array of strings or multiple arguments as keys for which EnvironmentPlugin injects process.env[key]. If a value of a key is undefined a warning is 
If a process.env variable added by webpack.EnvironmentPlugin is not defined, it is reported as a warning in the terminal.

**What is the new behavior?**
EnvironmentPlugin can now accept an object which uses the values as default values if none are set by process.env. This also allows the developer to silence the warnings by passing null as default value.

**Does this PR introduce a breaking change?**
- [x] No

einarlove commented:
I have not been able to write a test for this since I was unable to understand the current tests and even the current ones failed on `npm test`. Looked for documentation on how tests are written in webpack but couldn't find any. A helpful commit with a test is appreciated.

TheLarkInn commented:
To run tests you must run first: 

npm link && npm link webpack first

einarlove commented:
I did that after a lot of debugging. That should be documented for contributors.
My issue was that npm test took to long for me to iterate on the tests and I could not figure out how to only tests those specific files. I Have no previous experience with mocha.

Plus it as not immediately clear to me how [these tests](https://github.com/webpack/webpack/blob/master/test/configCases/plugins/environment-plugin/index.js) works. 😕

TheLarkInn commented:
> I did that after a lot of debugging. It should be documented for contributors.

https://github.com/webpack/webpack/blob/master/CONTRIBUTING.md We do :wink: 

> My issue was that npm test took to long for me to iterate on the tests and I could not figure out how to only tests those specific files. I Have no previous experience with mocha.

No worries! Instead I believe you can just run `mocha --harmony --check-leaks` and can specify  a specific file. The details for the mocha cli are here if you need some more help. https://mochajs.org/

sokra commented:
An idea:

Instead of allowing to silence warnings, we allow to pass default values to the plugin, which are used if no value is specified.

``` js
new EnvPlugin({
  NODE_ENV: "development", // default value, no warning if missing
  API_TOKEN: null, // no default value, warning if missing
  DEBUG: true // non-string values are stringified ("true")
})
```

moretti commented:
@sokra I think it's a good idea, but in order not to break backwards compatibility with need to use something like lodash's `isPlainObject`.

sokra commented:
just use `Array.isArray`.

einarlove commented:
Your idea of allowing an object with defaults is a good idea, @sokra. I rewrote it to allow it, and added a deprecation notice if you use pass multiple string arguments.

I tried to write a test for this, but the tests for this plugin is confusing and hard to reason with. Could you help by adding a test that should fail on multiple arguments and checks correct values when passed defaults, @sokra?

einarlove commented:
Force-updated to fix wrong email in commits for CLA signing

sokra commented on lib/EnvironmentPlugin.js:
```diff
@@ -1,17 +1,25 @@
 
 function EnvironmentPlugin(keys) {
-	this.keys = Array.isArray(keys) ? keys : Array.prototype.slice.call(arguments);
+	if (typeof keys === 'string') {
```
The old syntax should be also allowed. This doesn't need to be a breaking change.

einarlove replied:
I'll get on it. Had issues with definePlugin not working as expected in the 2.x.x beta.

einarlove commented:
Added more info to the warning message in 557b723.

<img width="703" alt="screen shot 2016-12-21 at 10 31 23 pm" src="https://cloud.githubusercontent.com/assets/1495211/21408531/7540c822-c7cd-11e6-8db5-362e23c3743f.png">

If it goes against your design or guidelines for warnings I can reverse it. I find it more helpful and love error/warning messages that gives me info on how to fix it.

einarlove commented:
When this merges the option of passing an object with default values should be added to the docs.

sokra commented on lib/EnvironmentPlugin.js:
```diff
@@ -1,28 +1,49 @@
 			compiler.plugin("this-compilation", function(compilation) {
-				var error = new Error(key + " environment variable is undefined.");
+				var error = new Error(
+					'EnvironmentPlugin - ' + key + ' environment variable is undefined. \033[0m\n\n' +
```
no color codes in not CLI related stuff. Colors could be disabled or not allowed on stdout.

einarlove replied:
Fixed

sokra commented on lib/EnvironmentPlugin.js:
```diff
@@ -1,28 +1,49 @@
+		this.keys = keys;
+	}
+
+	else if (keys === Object(keys)) {
```
`key && typeof keys === "object"`

sokra commented on lib/EnvironmentPlugin.js:
```diff
@@ -1,28 +1,49 @@
 		}
 
-		definitions["process.env." + key] = value ? JSON.stringify(value) : "undefined";
+		defs["process.env." + key] = value ? JSON.stringify(value) : "undefined";
```
`value` could be false or empty string. `typeof value === "undefined ? JSON.stringify(value) : "undefined"`

sokra commented on lib/EnvironmentPlugin.js:
```diff
@@ -1,28 +1,49 @@
+				var error = new Error(
+					'EnvironmentPlugin - ' + key + ' environment variable is undefined. \033[0m\n\n' +
+					'You can pass an object with default values to suppress this warning. \n' +
+					'See http://webpack.github.io/docs/list-of-plugins.html#environmentplugin for example.'
```
Could you include a link to the new documentation site?

webpack.js.org

einarlove replied:
The page describing plugins seems incomplete https://webpack.js.org/configuration/plugins. I don't have a hashtag to use since there is no documentation for EnvironmentPlugin—or any other plugins—so I don't know what the best course of action is.
1. Link to webpack.js.org alone
2. Link to webpack.js.org/plugins/#environmentplugin // A hashtag not does not exist yet
3. Leave it as is.

sokra commented:
Please style the code according to the beautify rules. `npm run beautify` elsewise the linting CI will fail.

sokra commented:
Your latest commits seem to have incorrect author information. Check if email in commit is email in github.

einarlove commented:
Updated the commits with new author. Working from another machine this Christmas vacation, so didn't notice the outdated email.

TheLarkInn commented:
@einarlove would you mind syncing with master and rebasing?

einarlove commented:
Done!

einarlove commented:
Docs have been proposed in this PR https://github.com/webpack/webpack.js.org/pull/620

einarlove commented on lib/EnvironmentPlugin.js:
```diff
@@ -38,7 +38,7 @@ class EnvironmentPlugin {
 			}
 
-			defs[`process.env.${key}`] = value === undefined ? "undefined" : JSON.stringify(value);
+			defs[`process.env.${key}`] = typeof value === undefined ? "undefined" : JSON.stringify(value);
```
Should be changed to `typeof value === undefined` if to work properly.
Why change it from `value === undefined`?
